File: backend/app/schemas/citation_canvas.py
from datetime import datetime
from typing import List

from pydantic import BaseModel, ConfigDict

from app.schemas.paper import Paper


class CanvasItemCreate(BaseModel):
  paper_id: int
  x: float
  y: float


class CanvasItemPosition(BaseModel):
  paper_id: int
  x: float
  y: float


class CanvasItemPositionUpdate(BaseModel):
  x: float
  y: float


class CanvasBulkPositions(BaseModel):
  items: List[CanvasItemPosition]


class CanvasItem(BaseModel):
  model_config = ConfigDict(from_attributes=True)

  paper_id: int
  x: float
  y: float
  created_at: datetime
  updated_at: datetime
  paper: Paper


class CanvasEdge(BaseModel):
  source: int
  target: int


class CanvasResponse(BaseModel):
  items: List[CanvasItem]
  edges: List[CanvasEdge]

File: frontend-v2/src/lib/api/citationCanvas.ts
import { api } from './client';
import type { Paper } from './papers';

export interface CanvasItem {
  paper_id: number;
  x: number;
  y: number;
  created_at: string;
  updated_at: string;
  paper: Paper;
}

export interface CanvasEdge {
  source: number;
  target: number;
}

export interface CanvasResponse {
  items: CanvasItem[];
  edges: CanvasEdge[];
}

export interface CanvasItemPosition {
  paper_id: number;
  x: number;
  y: number;
}

export const citationCanvasApi = {
  get: (): Promise<CanvasResponse> => api.get<CanvasResponse>('/citation-canvas'),

  addItem: (paperId: number, x: number, y: number): Promise<CanvasItem> =>
    api.post<CanvasItem>('/citation-canvas/items', { paper_id: paperId, x, y }),

  updatePosition: (paperId: number, x: number, y: number): Promise<CanvasItem> =>
    api.patch<CanvasItem>(`/citation-canvas/items/${paperId}`, { x, y }),

  bulkUpdatePositions: (items: CanvasItemPosition[]): Promise<void> =>
    api.post<void>('/citation-canvas/positions', { items }),

  removeItem: (paperId: number): Promise<void> =>
    api.delete<void>(`/citation-canvas/items/${paperId}`),

  clear: (): Promise<void> => api.delete<void>('/citation-canvas'),
};
